### ISSUE TITLE
certbot-apache fails to parse files with comma in the filename

### ISSUE THREAD

**Issue by joohoi** [certbot/certbot#2983]
Original issue was: #2949

Apache plugin errors out when trying to read files with comma in the filename.

```
letest@letest:/etc/apache2/sites-available$ ls -la
total 48
drwxr-xr-x 2 root root 4096 May 13 11:35 .
drwxr-xr-x 8 root root 4096 May 13 11:29 ..
-rw-r--r-- 1 root root 1878 May 10 13:36 000-default-le-ssl.conf
-rw-r--r-- 1 root root 1472 May 10 00:30 000-default.conf
-rw-r--r-- 1 root root  255 May 10 13:51 1.conf,OLD.conf
```

Running ```letsencrypt --apache --staging``` results in:

```
letest@letest:~$ sudo letsencrypt --apache --staging
An unexpected error occurred:
RuntimeError: Error during match procedure!
Please see the logfiles in /var/log/letsencrypt for more details.
```
With stack trace in the log file:
```
2016-05-13 08:35:32,962:DEBUG:letsencrypt.main:Root logging level set at 30
2016-05-13 08:35:32,963:INFO:letsencrypt.main:Saving debug log to /var/log/letsencrypt/letsencrypt.log
2016-05-13 08:35:32,963:DEBUG:letsencrypt.main:letsencrypt version: 0.5.0
2016-05-13 08:35:32,963:DEBUG:letsencrypt.main:Arguments: ['--apache', '--staging']
2016-05-13 08:35:32,964:DEBUG:letsencrypt.main:Discovered plugins: PluginsRegistry(PluginEntryPoint#apache,PluginEntryPoint#webroot,PluginEntryPoint#null,PluginEntryPoint#manual,PluginEntryPoint#standalone)
2016-05-13 08:35:32,967:DEBUG:letsencrypt.plugins.selection:Requested authenticator apache and installer apache
2016-05-13 08:35:33,292:DEBUG:letsencrypt.main:Exiting abnormally:
Traceback (most recent call last):
  File "/usr/bin/letsencrypt", line 9, in <module>
    load_entry_point('letsencrypt==0.5.0', 'console_scripts', 'letsencrypt')()
  File "/usr/lib/python2.7/dist-packages/letsencrypt/main.py", line 692, in main
    return config.func(config, plugins)
  File "/usr/lib/python2.7/dist-packages/letsencrypt/main.py", line 448, in run
    installer, authenticator = plug_sel.choose_configurator_plugins(config, plugins, "run")
  File "/usr/lib/python2.7/dist-packages/letsencrypt/plugins/selection.py", line 184, in choose_configurator_plugins
    authenticator = installer = pick_configurator(config, req_inst, plugins)
  File "/usr/lib/python2.7/dist-packages/letsencrypt/plugins/selection.py", line 24, in pick_configurator
    (interfaces.IAuthenticator, interfaces.IInstaller))
  File "/usr/lib/python2.7/dist-packages/letsencrypt/plugins/selection.py", line 76, in pick_plugin
    verified.prepare()
  File "/usr/lib/python2.7/dist-packages/letsencrypt/plugins/disco.py", line 214, in prepare
    return [plugin_ep.prepare() for plugin_ep in self._plugins.itervalues()]
  File "/usr/lib/python2.7/dist-packages/letsencrypt/plugins/disco.py", line 104, in prepare
    self._initialized.prepare()
  File "/usr/lib/python2.7/dist-packages/letsencrypt_apache/configurator.py", line 177, in prepare
    self.vhosts = self.get_virtual_hosts()
  File "/usr/lib/python2.7/dist-packages/letsencrypt_apache/configurator.py", line 552, in get_virtual_hosts
    new_vhost = self._create_vhost(path)
  File "/usr/lib/python2.7/dist-packages/letsencrypt_apache/configurator.py", line 505, in _create_vhost
    args = self.aug.match(path + "/arg")
  File "/usr/lib/python2.7/dist-packages/augeas.py", line 312, in match
    raise RuntimeError("Error during match procedure!")
RuntimeError: Error during match procedure!
```

This looks like a bug in Augeas, so pinging @domcleal

**Comment by domcleal** [comment]
Yes, it is a bug in Augeas, but it's fixed in version 1.4.0: https://github.com/hercules-team/augeas/blob/release-1.5.0/NEWS#L149-L151

The aug_match call in ApacheConfigurator's `get_virtual_hosts` returns a list of paths which in older versions aren't escaped, so they're not safe to pass back into aug_match in `_create_vhost`. It should work on 1.4.0 or higher.

That said, I'm unsure if some of the certbot code would handle it OK, namely `get_file_path` which assumes it can just strip off the `/files` prefix and `/VirtualHost` suffix to get the filesystem path from the Augeas path. It should probably instead strip the suffix only, then call `aug.aug_get("/augeas%s/path" % vhost_path)` to get the real filesystem path.  Some more testing might be needed!

**Comment by pde** [comment]
Hmmm, we're going to have trouble pushing all of our users up to augeas 1.4.0, so it might be best to deploy a mitigation where we catch the error and say something helpful in that case?

**Comment by pde** [comment]
@SwartzCr do you have time to take this?

**Comment by SwartzCr** [comment]
hmm, how are we specifying augeas version currently?
also joohoi why are you running version 0.5.0?
@pde to clarify you want a fix that either tries to bump the augeas version but also catches cases where it misbehaves and prints a better error - additionally it seems like I should take @domcleal's advice and ad some check inside the mentioned functions to test for comma'd filepaths

**Comment by domcleal** [comment]
I don't think there's any requirement in certbot specifying an Augeas version, only python-augeas, so better error handling should suffice, which won't be triggered by users on a newer Augeas version (assuming it all works).

**Comment by joohoi** [comment]
That was the most recent version when I was investigating this.

**Comment by domcleal** [comment]
The most recent version of Augeas, or certbot?  The version of Augeas may vary depending on your distribution.

The match works fine for me on 1.4.0, I think I'm doing the same as get_virtual_hosts and _create_vhost:

<pre>
>>> from augeas import Augeas
>>> a = Augeas(root="/tmp/root")
>>> a.match("/files/etc/apache2/sites-available//*[label()=~regexp('VirtualHost')]")
[u'/files/etc/apache2/sites-available/1.conf\\,OLD.conf/VirtualHost']
>>> a.match(_[0] + "/arg")
[u'/files/etc/apache2/sites-available/1.conf\\,OLD.conf/VirtualHost/arg']
$ cat /tmp/root/etc/apache2/sites-available/1.conf,OLD.conf
<VirtualHost *:80>
  ServerName test.example.com
</VirtualHost>
</pre>

Note the extra escaping returned from the first match call, which isn't present in 1.2.0 (on Ubuntu 14.04):

<pre>
>>> a.match("/files/etc/apache2/sites-available//*[label()=~regexp('VirtualHost')]")
['/files/etc/apache2/sites-available/1.conf,OLD.conf/VirtualHost']
>>> a.match(_[0] + "/arg")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/dist-packages/augeas.py", line 312, in match
    raise RuntimeError("Error during match procedure!")
RuntimeError: Error during match procedure!
</pre>

There's not much you can (or should) do about the returned value from the aug.match apart from handling any errors when you next call `aug.match(_[0])` gracefully for users without the fixed version.

**Comment by cosimosimeoneTR** [comment]
Guys, just a reminder: files with "strange" extensions should ***not*** be processed!
Processing .OLD, .tmp, .bck, .whatever might produce serious harm! :-)
Just a reminder, ROCK ON!!!

**Comment by SwartzCr** [comment]
there is now a PR open for this, would love your feedback @joohoi

**Comment by spacelama** [comment]
I think I just spent all day banging my head up against this bug.  I have two sites in my debian apache setup, and only one of them was being temporarily munged when going through the cert challenge.  The files were:
/etc/apache2/sites-enabled> l
lrwxrwxrwx 1 root root 55 Apr 18 16:46 0-http,https.rather.puzzling.org.conf -> ../sites-available/0-http,https.rather.puzzling.org.conf
lrwxrwxrwx 1 root root 52 Nov 18 14:36 1-httphttps.angelahughes.org.conf -> ../sites-available/1-httphttps.angelahughes.org.conf

certbot version is debian 1.12.0-2, but I think it was the same behaviour when I temporarily upgraded to 1.25.0-1

Tracing it with strace suggested to me to rename 0-http,https.rather.puzzling.org.conf to 0-httphttps.angelahughes.org.conf and it worked.

When I googled for "comma filename apache", thinking perhaps there's a well documented rule that apache conffiles shouldn't have a comma in them, I came across this bug.  So contrary to #issuecomment-226197407, the bug isn't because I was using a dodgy extension - I just had a comma in the filename in a perfectly legal place.  I can't what version this bug was supposedly fixed in from the commit in #event-725006102, but surely 1.12 was newer if this was being dealt with back in 2016?
